Fixed #7732 -- Added support for connection pooling in oracle backend.

suraj-ora-2020 opened the pull request:
[Ticket- 7732](https://code.djangoproject.com/ticket/7732)

cjbj commented on docs/ref/databases.txt:
```diff
@@ -1105,6 +1105,49 @@ Example of a full DSN string::
+
+.. versionadded:: 5.2
+
+To use a connection pool with `oracledb`_, you can set ``"pool"`` in the
```
Suggested change:
```
To use a connection pool with `oracledb`_, set ``"pool"`` to ``True`` in the
:setting:`OPTIONS` of your database configuration. This uses the driver's
`create_pool()`_ default values::
```

suraj-ora-2020 replied:
Updated.

cjbj commented on docs/ref/databases.txt:
```diff
@@ -1105,6 +1105,49 @@ Example of a full DSN string::
+
+.. _`oracledb`: https://oracle.github.io/python-oracledb/
+
+You can also set ``"pool"`` in the :setting:`OPTIONS` part of your database 
```
Suggested change:

```
To pass custom parameters to the driver's `create_pool()`_ function, you can
alternatively set ``"pool"`` to be a dict::
```

suraj-ora-2020 replied:
Updated the doc.

cjbj commented on docs/releases/5.2.txt:
```diff
@@ -123,7 +123,8 @@ CSRF
 ~~~~~~~~~~~~~~~~~
 
-* ...
+* ``"pool"`` option is now supported in :setting:`OPTIONS` on Oracle Database
```
change "on Oracle Database" to "for Oracle Database"

suraj-ora-2020 replied:
Done.

suraj-ora-2020 commented:
Thanks @cjbj have updated the docs as per your input.

suraj-ora-2020 commented:
@sarahboyce  The PR has been reviewed, and all relevant tests have passed. I have also addressed all feedback provided during the review process. Please let me know if you need any further information or if there are any additional steps required before merging.
Thank you for your attention to this request!

sarahboyce commented:
I imagine some action is required based off this: https://github.com/django/django/pull/17834#issuecomment-2341686247

> We need to bump the minimum requirement version 2.0.0 as from this version it ensures all the connections are closed before closing the pool (see [this](https://python-oracledb.readthedocs.io/en/latest/release_notes.html#:~:text=Internal%20change%20to%20ensure%20that%20all%20connections%20in%20a%20pool%20have%20been%20closed%20gracefully%20before%20the%20pool%20is%20closed.)).
Also in 2.3.0 contains major re-work in connection pooling which solves some known issues (see [this](https://python-oracledb.readthedocs.io/en/latest/release_notes.html#:~:text=When%20calling%20ConnectionPool.acquire()%20or%20AsyncConnectionPool.acquire()%2C%20the%20connection%20pool%20mode%20oracledb.POOL_GETMODE_TIMEDWAIT%20now%20always%20honors%20the%20wait_timeout%20value%20and%20the%20connection%20request%20will%20not%20additionally%20be%20delayed%20by%20any%20internal%20network%20ping%20to%20the%20database)).
The minimum we need is 2.0.0 and best would be 2.3.0. We may as well use 2.4.1 as the minimum version to encourage people to update the driver.

I am not sure if we should create a `DatabaseFeatures` flag for example

suraj-ora-2020 commented:
@sarahboyce I have bumped up the minimum supported version of oracledb to 2.3.0

sarahboyce commented on tests/requirements/oracle.txt:
```diff
@@ -1 +1 @@
-oracledb >= 1.3.2
+oracledb >= 2.3.0
```
@suraj-ora-2020 there's more work required to bump the minimum version (see 9946f0b0d9356b55e819f861b31615fa5b548f99 as an example)

We would need to: 
- update the docs saying ```Version xxx or higher of the `oracledb`_ Python driver is required``` in `docs/ref/databases.txt`
- add a release note along the lines of ```* The minimum supported version of ``oracledb`` is increased from 1.3.2 to
  xxxx.```

suraj-ora-2020 replied:
Done.

sarahboyce commented on docs/releases/5.2.txt:
```diff
@@ -123,7 +123,12 @@ CSRF
 ~~~~~~~~~~~~~~~~~
 
-* ...
+* Django now supports `oracledb`_ version 2.3.0 or higher.
```
We should move this to the "Miscellaneous" backwards incompatible changes section

```
* The minimum supported version of `oracledb`_ is increased from 1.3.2 to
  2.3.0.
```

suraj-ora-2020 replied:
Updated. Thanks.

sarahboyce commented:
buildbot, test on oracle.

sarahboyce commented on docs/ref/databases.txt:
```diff
@@ -1105,6 +1105,48 @@ Example of a full DSN string::
+            },
+        },
+    }
+
+.. _`oracledb`: https://oracle.github.io/python-oracledb/
```
```suggestion
```
We can remove this as we already have an oracledb reference link on this page

suraj-ora-2020 replied:
Done

sarahboyce commented on tests/requirements/oracle.txt:
```diff
@@ -1 +1 @@
-oracledb >= 1.3.2
+oracledb >= 2.3.0
```
Sorry @felixxm I actually don't know whether there are rules around when we do these updates. 
Skimmed the [release notes](https://python-oracledb.readthedocs.io/en/latest/release_notes.html#oracledb-2-3-0-july-2024) for the supported Python versions - it seems ok. Anything that we should be wary of?

felixxm replied:
We do them when we have to :shrug: e.g. when we want to use a new feature etc. It supports the same Python and Oracle versions as we do, so it should be painless. :+1:

suraj-ora-2020 commented:
buildbot, test on oracle.

suraj-ora-2020 commented:
buildbot

suraj-ora-2020 commented:
buildbot, test on oracle.

suraj-ora-2020 commented:
@sarahboyce Since all tests have passed, are we ready to proceed with the merge?

sarahboyce commented:
I have rebased and pushed some minor tweaks to the commit message and docs. Please be careful when pushing :+1:

sarahboyce commented:
buildbot, test on oracle.

sarahboyce commented:
> @sarahboyce Since all tests have passed, are we ready to proceed with the merge?

To me, this looks good. :tada: 

But I am in no rush to merge this. I would love to see some "real life" testing of this (i.e. with a Django project), and some approvals/more eyes from other contributors. The feature freeze is January 15th, so we have some time :+1:

suraj-ora-2020 commented:
We have a created a blog post to understand more about connection pooling, its benefits and how to use it with the PR. Please do checkout this [blog](https://medium.com/@sharad-chandran/boost-your-application-performance-with-oracle-connection-pooling-in-django-4e8b997ab815) to know more.

sarahboyce commented:
buildbot, test on oracle.

sarahboyce commented on django/db/backends/oracle/features.py:
```diff
@@ -139,6 +139,25 @@ def django_test_skips(self):
                     },
                 }
             )
+        if self.connection.is_pool:
```
Some tests hang for me when running with pool as True, such as:
- `select_for_update.tests.SelectForUpdateTests.test_raw_lock_not_available`
- `select_for_update.tests.SelectForUpdateTests.test_nowait_raises_error_on_block`
- `backends.tests.ThreadTests.test_default_connection_thread_local`

Can you check why?

suraj-ora-2020 replied:
After debugging the test `select_for_update.tests.SelectForUpdateTests.test_raw_lock_not_available`, I found that the test requires three connections. It attempts to acquire three connections without releasing them back to the pool. Since you have set `pool: True`, it uses the driver's default values, which set the maximum pool size to 2. Consequently, the pool becomes full, and the third acquire request hangs. Same for the other two tests.

While running the test suite, you can adjust the configuration to `'OPTIONS': {'pool': {'max': 3}}`. Alternatively, you could increase the maximum value even further, to 5 or 10. This change will fix the problem.

suraj-ora-2020 replied:
@sarahboyce The other option is to set the "get" mode to FORCE which will create new connections anyway when the pool is full. So you can the configuration as `'OPTIONS':{'pool':{'getmode':oracledb.POOL_GETMODE_FORCEGET}},` . With this option even if max is 2 new connection will be created if required. Although I believe 1st will be the best option.

sarahboyce replied:
I see, thank you!
These are different defaults to the [postgres connection pooling](https://www.psycopg.org/psycopg3/docs/api/pool.html#psycopg_pool.ConnectionPool) (which has a timeout for example) but that's fine
I like that we don't have an "opinion" as to sensible defaults and don't repeat the documentation of the external library. So in short, I think this requires no change

suraj-ora-2020 replied:
Yes, the default values for PostgreSQL connection pooling parameters differ. We discussed what the defaults should be and decided to leave them up to the driver. This approach helps keep the documentation and code simpler and more aligned with any changes in PostgreSQL. You can view the discussion here https://github.com/django/django/pull/17834#discussion_r1738024240.
Thank You.

suraj-ora-2020 commented:
buildbot, test on oracle.

suraj-ora-2020 commented:
buildbot

suraj-ora-2020 commented:
The CI error appears to be a configuration issue, and I believe it may be resolved with a second run. Can we rerun the CI?

sarahboyce commented:
> The CI error appears to be a configuration issue, and I believe it may be resolved with a second run. Can we rerun the CI?

Don't worry about this :+1:
